### ISSUE TITLE
MPT table API for proof of non-existance

### ISSUE THREAD

**Issue by CPerezz** [privacy-scaling-explorations/zkevm-specs#249]
> To be honest I'm not sure about:
> 
> > necessary to constrain "If the given account doesn't exist, then it will push 0 onto the stack instead." in the circuit.
> > And is there a method (could be used in circuit) to lookup if account is existing?
> 
> This got me thinking. I'd say yes, we definitely need to do this check right? But how? How can we assert that an account exists? This would require a lookup and I think we don't have any mechanism in our tables (see https://github.com/privacy-scaling-explorations/zkevm-specs/blob/master/specs/tables.md) that allows us to gather this info.
> 
> We would need to have a table with all the accounts for which we need to check it's existance. Or at the very least, to add a tag for this thing in the rw table for example. cc: @han0110 @ed255

Here's a summary of a discussion we had about this via chat:

There are 2 options for the MPT table API.
A. One is that you just do a balance lookup, and the MPT circuit constraints the value to be 0 when the account doesn't exist (using an MPT proof of non-existence).
B. The other option is to have a tag for the MPT table for account exists, which would mean that the BALANCE opcode needs to do 2 lookups:
1. Account exists
2. Account balance
I think option B is simpler for the MPT circuit, but I'm not sure if it already supports option A.

Some circuits require proving that an account doesn't exist.  With option B that is done with a single lookup.  With option A that would be done with the check `nonce == 0 && balance == 0 && code_hash == empty_hash`.

I'm more favorable towards option B.

For this PR I would say to go with one of the options (probably option A is simpler), and then we can revisit all lookups to non-existing accounts if necessary once we settle on which option we take.

_Originally posted by @ed255 in https://github.com/privacy-scaling-explorations/zkevm-specs/issues/248#issuecomment-1217937108_

**Comment by miha-stopar** [comment]
Agree, option B is easier for MPT and is already implemented.

**Comment by miha-stopar** [comment]
> So therefore that means that opcodes like `BALANCE` are pending from the MPT merge right?

Yes, but this also holds for state circuit.

**Comment by ed255** [comment]
Depends on https://github.com/privacy-scaling-explorations/zkevm-specs/pull/236 which updates the MPT table spec.

**Comment by miha-stopar** [comment]
Currently, to reduce the expression degree MPT uses multiple boolean columns to determine what kind of proof it is about.

So, instead of
```
| Address | FieldTag | Key | ValuePrev | Value | RootPrev | Root |
| - | - | - | - | - | - | - |
| $addr | Nonce | 0 | $noncePrev | $nonceCur | $rootPrev | $root |
```

there is
```
| Address | IsStorageMod | IsNonceMod | IsBalanceMod | IsCodeHashProof | IsAccountDeleteMod | IsNonExistingAccountProof | Key | ValuePrev | Value | RootPrev | Root |
| - | - | - | - | - | - | - |
| $addr | 0 | 1 | 0 | 0 | 0 | 0 | 0 | $noncePrev | $nonceCur | $rootPrev | $root |
```

For `IsAccountDeleteMod` and `IsNonExistingAccountProof` we use 0 for `ValuePrev` and `Value`. Note that there is no `AccountCreateProof` because we can create the account implicitly by `IsNonceMod` or `IsBalanceMod` (and in this case we don't need to check that the account previously didn't exist, while for `IsAccountDeleteMod` the MPT circuit checks whether there really is not an account with the specified address anymore).

Of course we can change this if the number of columns becomes an issue. If it is ok having multiple selector columns, I can prepare update specs to reflect that. What do you think?

**Comment by ed255** [comment]
> Regarding the questions - yes to all, except for the last one. If there is no storage leaf neither before or after, currently MPT circuit cannot prove anything. But do we need this case?

This case can definitely happen.  Imagine a storage slot has value 0, and then we find an SSTORE that writes 0 into it.  Or imagine a storage slot has value 0 and then we do a read to it with SLOAD.
I think we have 2 solutions: handle this case from the MPT circuit, or handle it from the EVM circuit.  
- a. From the MPT side it would mean that a storage update with `value_prev=0, value=0` means that the MPT proof constraints are disabled somehow.
- b. From the EVM side it would mean that a read of value 0, or a write with `value_prev=0, value=0` don't trigger a lookup (so we'll need to use the IsZeroGadget to and get a dynamic selector expression to disable the lookup)

I don't have a preference for any option.  I'm not aware of how complex option a) is, and if it increases the MPT circuit cost.  Option b) seems simple and I don't think it increases the EVM circuit cost.

**Comment by miha-stopar** [comment]
> This case can definitely happen. Imagine a storage slot has value 0, and then we find an SSTORE that writes 0 into it. Or imagine a storage slot has value 0 and then we do a read to it with SLOAD.

Yes, but the first case means that there is a storage leaf in the trie after modification, the second means that there is a storage leaf in the trie before the modification? If that is so, both cases are covered by MPT circuit.

**Comment by ed255** [comment]
> > This case can definitely happen. Imagine a storage slot has value 0, and then we find an SSTORE that writes 0 into it. Or imagine a storage slot has value 0 and then we do a read to it with SLOAD.
> 
> Yes, but the first case means that there is a storage leaf in the trie after modification, the second means that there is a storage leaf in the trie before the modification? If that is so, both cases are covered by MPT circuit.

I don't think so.  In the first case, the leaf doesn't exist before, and since you write a 0, it remains non-existing, because a 0-value leaf in the storage will never exist.
In the second case there's no leaf, and a read to a key that has no leaf returns the value 0.
Here's the reference in go-ethereum:
- Writing a 0 value to storage is a delete: https://github.com/ethereum/go-ethereum/blob/389021a5afd01147c851870c693ded8760e6a08c/core/state/state_object.go#L333

But I realize now that there's a difference between the MPT spec, and the EVM storage behavior.  What I described is the EVM account storage behaviour.  So maybe from the MPT side, there can be leafs with the 0 value, but from the EVM side, a storage slot with value 0 means that the leaf in the MPT doesn't exist.  And we need to figure out in which side we implement this extra logic.

**Comment by ed255** [comment]
> True, totally forgot about that :)
> 
> But it's actually implemented this way in MPT circuit too (leaf is checked to be deleted, not just value set to 0). So it's possible to implement the above-mentioned case in MPT too, but if it is easy to handle this in the EVM circuit (b option), I would prefer that way.

Great!  Implementing this case in the EVM circuit sounds good to me!  I think it should be easy, and probably we're already doing checks against 0 values for variable gas cost purposes.

**Comment by ed255** [comment]
Hm, after a second thought... I believe we can't just skip that lookup from the EVM side, but instead we need some form of `IsNonExistingStorageProof`.  Otherwise the prover could just pretend to read a 0 value when there was a non-zero value in the storage.

**Comment by miha-stopar** [comment]
I see, I will add that.

**Comment by ed255** [comment]
Tasks to do:
- Add `IsNonExistingStorageProof` to the MPT table in `tables.md`
- Figure out a way to introduce `Account->NonExistingAccountProof` and `Storage->IsNonExistingStorageProof` in the RwTable
  - Update `tables.md` and python implementation accordingly
